pkg/dm/spec/topology_dm.go

```diff
@@ -0,0 +1,742 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spec
+
+import (
+	"fmt"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/creasty/defaults"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tiup/pkg/cluster/api"
+	"github.com/pingcap/tiup/pkg/cluster/clusterutil"
+	"github.com/pingcap/tiup/pkg/cluster/spec"
+	"github.com/pingcap/tiup/pkg/meta"
+	"github.com/pingcap/tiup/pkg/set"
+)
+
+const (
+	statusQueryTimeout = 2 * time.Second
+)
+
+var (
+	globalOptionTypeName  = reflect.TypeOf(GlobalOptions{}).Name()
+	serverConfigsTypeName = reflect.TypeOf(DMServerConfigs{}).Name()
+)
+
+func setDefaultDir(parent, role, port string, field reflect.Value) {
+	if field.String() != "" {
+		return
+	}
+	if defaults.CanUpdate(field.Interface()) {
+		dir := fmt.Sprintf("%s-%s", role, port)
+		field.Set(reflect.ValueOf(filepath.Join(parent, dir)))
+	}
+}
+
+func findField(v reflect.Value, fieldName string) (int, bool) {
+	for i := 0; i < v.NumField(); i++ {
+		if v.Type().Field(i).Name == fieldName {
+			return i, true
+		}
+	}
+	return -1, false
+}
+
+// Skip global/monitored/job options
+func isSkipField(field reflect.Value) bool {
+	tp := field.Type().Name()
+	return tp == globalOptionTypeName || tp == serverConfigsTypeName
+}
+
+type (
+	// GlobalOptions of spec.
+	GlobalOptions = spec.GlobalOptions
+	// PrometheusSpec is the spec of Prometheus
+	PrometheusSpec = spec.PrometheusSpec
+	// GrafanaSpec is the spec of Grafana
+	GrafanaSpec = spec.GrafanaSpec
+	// AlertManagerSpec is the spec of Alertmanager
+	AlertManagerSpec = spec.AlertManagerSpec
+	// ResourceControl is the spec of ResourceControl
+	ResourceControl = meta.ResourceControl
+)
+
+type (
+	// DMServerConfigs represents the server runtime configuration
+	DMServerConfigs struct {
+		Master map[string]interface{} `yaml:"dm-master"`
+		Worker map[string]interface{} `yaml:"dm-worker"`
+		Portal map[string]interface{} `yaml:"dm-portal"`
+	}
+
+	// DMTopologySpecification represents the specification of topology.yaml
+	DMTopologySpecification struct {
+		GlobalOptions GlobalOptions `yaml:"global,omitempty"`
+		// MonitoredOptions MonitoredOptions   `yaml:"monitored,omitempty"`
+		ServerConfigs DMServerConfigs    `yaml:"server_configs,omitempty"`
+		Masters       []MasterSpec       `yaml:"dm-master_servers"`
```

Comment:
Mix the `_` and `-` is very weird.

Reply:
https://github.com/pingcap/tiup/pull/554/commits/2788cbde31bcba797d9a98a99bff86f7bfd8cb89